pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	if err := st.Merge(p.Data); err != nil {
+		level.Error(s.logger).Log("msg", "failed to merge state", "err", err, "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return errors.Wrap(err, "unable to merge state")
+	}
+
+	return nil
+}
+
+// The clusterWait function uses the position to determine how long should it wait before sending a notification.
+// In the Cortex case, we copy the function directly and use the manager to figure out which position we hold.
+func (s *state) Position() int { return 0 }
+
+// Settle waits until the alertmanagers are ready (and sets the appropriate internal state when it is).
+// The idea is that we don't want to start working" before we get a chance to know most of the notifications and/or silences.
+func (s *state) Settle(ctx context.Context, _ time.Duration) {
+	level.Info(s.logger).Log("msg", "Waiting for notification and silences to settle...")
+
+	// TODO: Make sure that the state is fully synchronised at this point.
+	// We can check other alertmanager(s) and explicitly ask them to propagate their state to us if available.
+	close(s.readyc)
+}
+
+// WaitReady is needed for the pipeline builder to know whenever we've settled and the state is up to date.
+func (s *state) WaitReady() {
+	//TODO: At the moment, we settle in a separate go-routine (see multitenant.go as we create the Peer) we should
+	// mimic that behaviour here once we have full state replication.
+	s.Settle(context.Background(), time.Second)
+	<-s.readyc
+}
+
+func (s *state) Ready() bool {
+	select {
+	case <-s.readyc:
+		return true
+	default:
+	}
+	return false
+}
+
+func (s *state) handleBroadcasts() {
+	for {
+		select {
+		case p := <-s.msgc:
+			// If the replication factor is <= 1, we don't need to replicate any state anywhere else.
+			if s.replicationFactor <= 1 {
+				return
+			}
+
+			s.stateReplicationTotal.WithLabelValues(p.Key).Inc()
+			ctx := context.Background() //TODO: I probably need a better context
+			if err := s.replicateStateFunc(ctx, s.userID, p); err != nil {
+				s.stateReplicationFailed.WithLabelValues(p.Key).Inc()
+				level.Error(s.logger).Log("msg", "failed to replicate state to other alertmanagers", "user", s.userID, "key", p.Key, "err", err)
+			}
+		case <-s.stopc:
+			return
+		}
+	}
+}
+
+// stateChannel allows a state publisher to send messages that will be broadcasted to all other alertmanagers that a tenant
+// belongs to.
+type stateChannel struct {
+	key    string
+	logger log.Logger
+
+	send func([]byte)
+}
+
+func newStateChannel(key string, send func([]byte), logger log.Logger) *stateChannel {
+	c := &stateChannel{
+		key:    key,
+		logger: logger,
+		send:   send,
+	}
+
+	return c
+}
+
+func (c *stateChannel) Broadcast(b []byte) {
+	//TODO: We probably need a bounded queue at some point to avoid a single tenant overflowing with messages
+	level.Debug(c.logger).Log("msg", "message received for broadcasting", "key", c.key)
```

Comment:
We can move this logging into `MultitenantAlertmanager.replicateStateForUser` and simplify this struct.